File: testdata/bench/bench.java
package testdata.bench;

import org.junit.Test;

import java.util.Arrays;


public class bench {

	private static Colfer[] newTestData() {
		Colfer c1 = new Colfer();
		c1.key = 1234567890L;
		c1.host = "db003lz12";
		c1.port = 389;
		c1.size = 452;
		c1.hash = 0x488b5c2428488918L;
		c1.ratio = 0.99;
		c1.route = true;

		Colfer c2 = new Colfer();
		c2.key = 1234567891L;
		c2.host = "localhost";
		c2.port = 22;
		c2.size = 4096;
		c2.hash = 0x243048899c24c824L;
		c2.ratio = 0.20;
		c2.route = false;

		Colfer c3 = new Colfer();
		c3.key = 1234567892L;
		c3.host = "kdc.local";
		c3.port = 88;
		c3.size = 1984;
		c3.hash = 0x000048891c24485cL;
		c3.ratio = 0.06;
		c3.route = false;

		Colfer c4 = new Colfer();
		c4.key = 1234567893L;
		c4.host = "vhost8.dmz.example.com";
		c4.port = 27017;
		c4.size = 59741;
		c4.hash = 0x5c2408488b9c2489L;
		c4.ratio = 0.0;
		c4.route = true;

		return new Colfer[] {c1, c2, c3, c4};
	}

	// prevent compiler optimization
	public static byte[] holdSerial;
	public static Colfer holdData;

	@Test
	public void benchMarshal() {
		Colfer[] testData = newTestData();
		final int n = 20000000;

		long start = System.nanoTime();
		for (int i = 0; i < n; i++) {
			holdSerial = new byte[200];
			testData[i % testData.length].marshal(holdSerial, 0);
		}
		long end = System.nanoTime();

		System.err.printf("%dM marshals avg %dns\n", n / 1000000, (end - start) / n);
	}

	@Test
	public void benchMarshalReuse() {
		Colfer[] testData = newTestData();
		holdSerial = new byte[Colfer.colferSizeMax];
		final int n = 20000000;

		long start = System.nanoTime();
		for (int i = 0; i < n; i++) {
			testData[i % testData.length].marshal(holdSerial, 0);
		}
		long end = System.nanoTime();

		System.err.printf("%dM marshals with buffer reuse avg %dns\n", n / 1000000, (end - start) / n);
	}

	@Test
	public void benchUnmarshal() {
		Colfer[] testData = newTestData();
		byte[][] serials = new byte[testData.length][];
		for (int i = 0; i < serials.length; i++) {
			byte[] buf = new byte[200];
			int n = testData[i].marshal(buf, 0);
			serials[i] = Arrays.copyOf(buf, n);
		}
		final int n = 20000000;

		long start = System.nanoTime();
		for (int i = 0; i < n; i++) {
			holdData = new Colfer();
			holdData.unmarshal(serials[i % serials.length], 0);
		}
		long end = System.nanoTime();

		System.err.printf("%dM unmarshals avg %dns\n", n / 1000000, (end - start) / n);
	}

}

File: ecma/bench/bench.js
global._ = require('./lodash.js')
const Benchmark = require('./benchmark.js');

const Colfer = require('./build/Colfer.js')

const testData = [
	new Colfer.bench.Colfer({key: 1234567890, host: "db003lz12", port: 389, size: 452, hash: 0x5c2428488918, ratio: 0.99, route: true}),
	new Colfer.bench.Colfer({key: 1234567891, host: "localhost", port: 22, size: 4096, hash: 0x48899c24c824, ratio: 0.20, route: false}),
	new Colfer.bench.Colfer({key: 1234567892, host: "kdc.local", port: 88, size: 1984, hash: 0x48891c24485c, ratio: 0.06, route: false}),
	new Colfer.bench.Colfer({key: 1234567893, host: "vhost8.dmz.example.com", port: 27017, size: 59741, hash: 0x08488b9c2489, ratio: 0.0, route: true}),
];

// Corresponding testData Colfer serials.
var testColfer = new Array(testData.length);
testData.forEach(function(o, i) {
	testColfer[i] = o.marshal();
});

// Corresponding testData JSON serials.
var testJSON = new Array(testData.length);
testData.forEach(function(o, i) {
	testJSON[i] = JSON.stringify(o);
});

var suite = new Benchmark.Suite;
suite.add('marshal Colfer', function() {
		testData[0].marshal();
		testData[1].marshal();
		testData[2].marshal();
		testData[3].marshal();
	})
	.add('unmarshal Colfer', function() {
		new Colfer.bench.Colfer().unmarshal(testColfer[0]);
		new Colfer.bench.Colfer().unmarshal(testColfer[1]);
		new Colfer.bench.Colfer().unmarshal(testColfer[2]);
		new Colfer.bench.Colfer().unmarshal(testColfer[3]);
	})
	.add('marshal JSON', function() {
		JSON.stringify(testData[0]);
		JSON.stringify(testData[1]);
		JSON.stringify(testData[2]);
		JSON.stringify(testData[3]);
	})
	.add('unmarshal JSON', function() {
		JSON.parse(testJSON[0]);
		JSON.parse(testJSON[1]);
		JSON.parse(testJSON[2]);
		JSON.parse(testJSON[3]);
	})
	.on('error', function(event) {
		console.log(String(event.target.error));
	})
	.on('cycle', function(event) {
		console.log(String(event.target))
	})
	.run({async: true});
